### PR TITLE
Unify physics modules and test names

### PR DESCRIPTION
**Pull request by jamiebramwell**
- Renames the old physics modules to `Legacy` and removes the extra `Functional` from the new modules
- Adds optional non-default mesh field to the new `Functional`-based tests
- Renames the unit tests to remove the extra serac prefix
- Renames modules to `SolidMechanics`, `Thermomechanics`, and `HeatTransfer`

### CODE REVIEW

File: src/serac/physics/solid.hpp [hunk 1]
```diff
@@ -7,631 +7,907 @@
...
+   * @param qfunction a callable that returns a tuple of body-force and stress
+   * @param qdata the buffer of material internal variables at each quadrature point
    *
-   * @param[in] ext_force_coef Add a vector-valued external force coefficient applied to the domain
-   */
-  void addBodyForce(std::shared_ptr<mfem::VectorCoefficient> ext_force_coef);
-
-  /**
-   * @brief Set the viscosity coefficient
+   * ~~~ {.cpp}
```
Comment (white238) [inline src/serac/physics/solid.hpp]: Outside of this PR but what is this?

Comment (white238) [reply src/serac/physics/solid.hpp]: oh its a code snippet.. interesting.

File: src/serac/physics/solid_legacy.hpp [hunk 2]
```diff
@@ -0,0 +1,638 @@
+// Copyright (c) 2019, Lawrence Livermore National Security, LLC and
```
Comment (white238) [inline src/serac/physics/solid_legacy.hpp]:
```suggestion
// Copyright (c) 2019-2022, Lawrence Livermore National Security, LLC and
```

File: src/serac/physics/tests/CMakeLists.txt [hunk 3]
```diff
@@ -15,10 +15,10 @@ blt_add_library(
...
 
 set(serial_solver_tests
-    serac_solid_sensitivity.cpp
-    serac_thermal_functional_finite_diff.cpp
-    serac_solid_functional_finite_diff.cpp
-    test_solid_functional_statics.cpp
+    solid_legacy_sensitivity.cpp
+    thermal_finite_diff.cpp
+    solid_finite_diff.cpp
+    test_solid_statics.cpp
```
Comment (white238) [inline src/serac/physics/tests/CMakeLists.txt]: does this need to be prefixed with test?

Comment (jamiebramwell) [reply src/serac/physics/tests/CMakeLists.txt]: probably not

File: src/serac/physics/thermal_mechanics.hpp [hunk 4]
```diff
@@ -37,13 +37,14 @@ class ThermalMechanicsFunctional : public BasePhysics {
...
-                             GeometricNonlinearities geom_nonlin = GeometricNonlinearities::On,
-                             const std::string&      name        = "")
-      : BasePhysics(3, order, name),
-        thermal_functional_(thermal_options, name + "thermal"),
-        solid_functional_(solid_options, geom_nonlin, name + "mechanical")
+  ThermalMechanics(const SolverOptions& thermal_options, const SolverOptions& solid_options,
+                   GeometricNonlinearities geom_nonlin = GeometricNonlinearities::On, const std::string& name = "",
+                   mfem::ParMesh* pmesh = nullptr)
+      : BasePhysics(3, order, name, pmesh),
+        thermal_functional_(thermal_options, name + "thermal", pmesh),
```
Comment (white238) [inline src/serac/physics/thermal_mechanics.hpp]: This should be renamed

File: src/serac/physics/thermal_mechanics.hpp [hunk 5]
```diff
@@ -15,8 +15,8 @@
 #include "mfem.hpp"
 
 #include "serac/physics/base_physics.hpp"
-#include "serac/physics/solid_functional.hpp"
-#include "serac/physics/thermal_conduction_functional.hpp"
+#include "serac/physics/solid.hpp"
+#include "serac/physics/thermal_conduction.hpp"
 #include "serac/physics/materials/thermal_functional_material.hpp"
 #include "serac/physics/materials/solid_functional_material.hpp"
```
Comment (white238) [inline src/serac/physics/thermal_mechanics.hpp]: These headers as well

File: src/serac/physics/tests/solid_statics_patch.cpp [hunk 6]
```diff
@@ -186,7 +186,7 @@ double solution_error(const ExactSolution& exact_displacement, PatchBoundaryCond
   auto solver_options = direct_static_options;
   solver_options.nonlinear.abs_tol = 1e-14;
   solver_options.nonlinear.rel_tol = 1e-14;
-  SolidFunctional<p, dim> solid_functional(solver_options, GeometricNonlinearities::On, "solid_functional");
+  Solid<p, dim> solid_functional(solver_options, GeometricNonlinearities::On, "solid_functional");
```
Comment (white238) [inline src/serac/physics/tests/solid_statics_patch.cpp]: rename this

File: src/serac/physics/thermal_solid_legacy.hpp [hunk 7]
```diff
@@ -5,7 +5,7 @@
 // SPDX-License-Identifier: (BSD-3-Clause)
```
Comment (white238) [inline src/serac/physics/thermal_solid_legacy.hpp:5]: This copy right is wrong.. can you grep the code for "Copyright (c) 2019," there seems to be a lot of these slipping through or search/replaces.

**Comment by samuelpmishLLNL** [comment]
I apologize in advance for bike-shedding, but I don't think it's avoidable in a PR like this. Some thoughts:

- I like the name `SolidMechanics` for that module better than `Solid`
- `ThermalConduction` sounds too specific, users can specify other mechanisms of heat transfer (e.g. Newton's law of cooling, radiative emission, etc) in this module. I'm not sure what a good name would be though. `HeatTransfer`?
- Similarly, `ThermalSolid` vs `Thermomechanics` vs. `ThermalMechanics` (I'd vote for `Thermomechanics`).

**Review by white238 — APPROVED** [review]

**Comment by btalamini** [comment]
As Sam pointed out, this discussion has the potential to go way off the rails, so close or delete this if necessary. Anyway, here is my name critique:

> I apologize in advance for bike-shedding, but I don't think it's avoidable in a PR like this. Some thoughts:
> 
>     * I like the name `SolidMechanics` for that module better than `Solid`

+1 for `SolidMechanics`.

>     * `ThermalConduction` sounds too specific, users can specify other mechanisms of heat transfer (e.g. Newton's law of cooling, radiative emission, etc) in this module. I'm not sure what a good name would be though. `HeatTransfer`?

Agree that `HeatTransfer` is better then `ThermalConduction` for the same reason. It's probably also the most idiomatic term for it. OTOH, we might also want to think how the names compose for multi-physics. Either of `ThermalTransport` or `ThermalDiffusion` gets the word "thermal" in there, which helps explain what  we mean by `Thermomechanics` (or `ThermalMechanics` if we go that way).
 
>     * Similarly, `ThermalSolid` vs `Thermomechanics` vs. `ThermalMechanics` (I'd vote for `Thermomechanics`).

+1 for `Thermomechanics`. Though worth noting that this naming convention does get cumbersome when the number of physics increases (eg `Thermochemomechanics`).

**Comment by white238** [comment]
Is there anything that needs to be updated in `src/docs`?

**Comment by jamiebramwell** [comment]
I think it's fine to break the composable names and `HeatTransfer` is a much more accurate description. Also this is a great PR for bikeshedding as I'm touching almost every file anyway. 😆

**Review by samuelpmishLLNL — APPROVED** [review]

**Pull request merged**